### PR TITLE
Subscribing Zhihu topics.

### PR DESCRIPTION
**Pull request by sschyu**
Addressing https://github.com/lilydjwg/morerssplz/issues/31

### CODE REVIEW

File: morerss/zhihu_stream.py [hunk 1]
```diff
@@ -69,6 +100,30 @@ class ZhihuAPI:
...
+    """
+    intro_url = 'https://www.zhihu.com/api/v4/topics/%s/intro' % name
+    res = await fetch_zhihu(
+      intro_url, headers={'User-Agent': self.user_agent})
+    if not res.body:
+      # e.g. https://www.zhihu.com/bei-feng-san-dai
+      raise web.HTTPError(404)
+    intro = json.loads(res.body.decode('utf-8'))
+    # topic name
+    name = intro['content']['meta']['name']
```
Comment (lilydjwg) [inline morerss/zhihu_stream.py]: I got `KeyError: 'meta'` on this line, e.g. https://www.zhihu.com/api/v4/topics/19667817/intro doesn't have a meta key.

Comment (sschyu) [reply morerss/zhihu_stream.py]: Zhihu's API has some weird inconsistencies. :frowning_face:  I'll find a workaround.

**Comment by sschyu** [comment]
Some Zhihu topics just don't have `content.meta.name` field in their `/api/v4/topics/{topic-id}/intro` APIs. I think the workaround is to scrape the topic's title and description from the page using XPath.
I'm working on this.

**Comment by sschyu** [comment]
The weirder thing is that the data from `/api/v4/topics/{topic-id}/feeds/top_activity` response isn't in line with what is displayed on the web page.

File: morerss/zhihu_stream.py (hunk 1 above)
Comment (sschyu) [reply morerss/zhihu_stream.py]: https://github.com/lilydjwg/morerssplz/pull/32/commits/6d01a6e1a23ba5f68628e03cc52a522b276f67c8 fixes the missing topic name.

File: morerss/zhihu_stream.py [hunk 2]
```diff
@@ -122,14 +182,21 @@ def post2rss(post, digest=False, pic=None):
...
+  if 'content' not in post or digest:
     content = post['excerpt']
   else:
     content = post['content']
 
   content = content.replace('<code ', '<pre><code ')
   content = content.replace('</code>', '</code></pre>')
 
+  # Post only contains images but no text
+  if not content:
```
Comment (SiqingYu) [inline morerss/zhihu_stream.py]: https://github.com/lilydjwg/morerssplz/pull/32/commits/6d01a6e1a23ba5f68628e03cc52a522b276f67c8 also addresses some posts that only contain images but no text in the Zhihu topics API.

File: morerss/zhihu_stream.py [hunk 3]
```diff
@@ -122,14 +182,21 @@ def post2rss(post, digest=False, pic=None):
...
+  # Posts in Zhihu topics API response don't have the 'content' key by default
+  # Although they can include it by carrying verbose query params
+  # which are hard to maintain because Zhihu doesn't have public API documentation :(
+  if 'content' not in post or digest:
     content = post['excerpt']
   else:
     content = post['content']
 
   content = content.replace('<code ', '<pre><code ')
   content = content.replace('</code>', '</code></pre>')
```
Comment (SiqingYu) [inline morerss/zhihu_stream.py]: 6d01a6e also addresses some posts that only contain images but no text in the Zhihu topics API.

File: morerss/zhihu_stream.py [hunk 4]
```diff
@@ -122,14 +182,21 @@ def post2rss(post, digest=False, pic=None):
...
 
+  # Post only contains images but no text
+  if not content:
+    content = '<img src="%s">' % post.get('thumbnail')
+
   doc = fromstring(content)
```
Comment (sschyu) [inline morerss/zhihu_stream.py]: 6d01a6e also addresses some posts that only contain images but no text in the Zhihu topics API.

**Comment by sschyu** [comment]
Oh. I finally figure out multi-line commenting. GitHub is awesome. :tada: 

https://twitter.com/github/status/1179101186437324801

File: morerss/zhihu_stream.py [hunk 5]
```diff
@@ -100,27 +100,32 @@ class ZhihuAPI:
...
+    elif doc.xpath('//h1[@class="TopicCard-titleText"]'):
+      name = doc.xpath('//h1[@class="TopicCard-titleText"]')[0].text_content()
+    else:
+      name = id
+
     return {
       'name': name,
-      'description': description,
+      # Omit the description
+      'description': '',
```
Comment (lilydjwg) [inline morerss/zhihu_stream.py]: Why is it always omitted? If it's not always available, do not use it.

File: morerss/zhihu_stream.py [hunk 6]
```diff
@@ -252,20 +262,20 @@ class ZhihuStream(base.BaseHandler):
     self.finish(rss)
 
 class ZhihuTopic(base.BaseHandler):
-  async def get(self, name):
+  async def get(self, id):
     """
-    :param name (str): Zhihu topic slug, as "19551894" in "https://www.zhihu.com/topic/19551894/hot"
+    :param id (str): Zhihu topic slug, as "19551894" in "https://www.zhihu.com/topic/19551894/hot"
```
Comment (lilydjwg) [inline morerss/zhihu_stream.py]: It's an id, not a slug. No doc is better than wrong doc....

**Review by lilydjwg — CHANGES_REQUESTED** [review]
Also, you'll need to include processing of type=question in post2rss function.

File: morerss/zhihu_stream.py (hunk 6 above)
Comment (sschyu) [reply morerss/zhihu_stream.py]: f669c1a fixes the term error.

File: morerss/zhihu_stream.py (hunk 5 above)
Comment (sschyu) [reply morerss/zhihu_stream.py]: 637bfda adds `description` field.

File: morerss/zhihu_stream.py (hunk 6 above)
Comment (SiqingYu) [reply morerss/zhihu_stream.py]: > 就是知乎新版个人页面的动态啦。不过只支持回答和文章两种类型。
https://rss.lilydjwg.me/

@lilydjwg I thought you ignore the question type by design. If I modify `morerss.zhihu_stream.post2rss` to process the questions, then users feeds like https://rss.lilydjwg.me/zhihu/jian-dan-xin-li will include questions too.

Should `morerss.zhihu_stream.post2rss` include questions? Or should I make this for topics only?

**Comment by sschyu** [comment]
> 就是知乎新版个人页面的动态啦。不过只支持回答和文章两种类型。
https://rss.lilydjwg.me/

@lilydjwg I thought you ignored the question type by design. If I modify `morerss.zhihu_stream.post2rss` to process the questions, then users feeds like https://rss.lilydjwg.me/zhihu/jian-dan-xin-li will include questions too.

Should `morerss.zhihu_stream.post2rss` include questions? Or should I make this for topics only?

**Comment by lilydjwg** [comment]
Yes it's ignored by zhihu_stream by design because one subscribes to get information, not to answer questions or follow people's activities in general.

Please make it for topics only. You may use an additional argument like `extra_types` to distinguish.

**Comment by sschyu** [comment]
fdf657b adds support for question post type in topics.

**Comment by lilydjwg** [comment]
Thanks!

**Pull request merged**